senders/msteams/msteams.go

```diff
@@ -0,0 +1,240 @@
+	sender.frontURI = senderSettings["front_uri"]
+	maxEvents, err := strconv.Atoi(senderSettings["max_events"])
+	if err != nil {
+		return fmt.Errorf("max_events should be an integer: %w", err)
+	}
+	sender.maxEvents = maxEvents
+	sender.client = &http.Client{
+		Timeout: time.Duration(30) * time.Second,
+	}
+	return nil
+}
+
+// SendEvents implements Sender interface Send
+func (sender *Sender) SendEvents(events moira.NotificationEvents, contact moira.ContactData, trigger moira.TriggerData, plot []byte, throttled bool) error {
+
+	err := sender.isValidWebhookURL(contact.Value)
+	if err != nil {
+		return err
+	}
+
+	request, err := sender.buildRequest(events, contact, trigger, plot, throttled)
+
+	if err != nil {
+		return fmt.Errorf("failed to build request: %w", err)
+	}
+
+	response, err := sender.client.Do(request)
+
+	if err != nil {
+		return fmt.Errorf("failed to perform request: %w", err)
+	}
+	defer response.Body.Close()
+
+	// read the entire response as required by https://golang.org/pkg/net/http/#Client.Do
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return fmt.Errorf("failed to decode response: %w", err)
+	}
+
+	//handle non 2xx responses
+	if response.StatusCode >= http.StatusBadRequest && response.StatusCode <= http.StatusNetworkAuthenticationRequired {
+		return fmt.Errorf("server responded with a non 2xx code: %d", response.StatusCode)
+
+	}
+
+	responseData := string(body)
+	if responseData != teamsOKResponse {
+		return fmt.Errorf("teams endpoint responded with an error: %s", responseData)
+	}
+
+	return nil
+}
+
+func (sender *Sender) buildMessage(events moira.NotificationEvents, trigger moira.TriggerData, throttled bool) MessageCard {
+
+	title, uri := sender.buildTitleAndURI(events, trigger)
+	var triggerDescription string
+	if trigger.Desc != "" {
+		triggerDescription = string(blackfriday.Run([]byte(trigger.Desc)))
+	}
+	facts := sender.buildEventsFacts(events, sender.maxEvents, throttled)
+	var actions []Actions
+	if uri != "" {
+		actions = append(actions, Actions{
+			Type: openUri,
+			Name: openUriMessage,
+			Targets: []OpenURITarget{
+				{
+					Os:  openUriOsDefault,
+					URI: uri,
+				},
+			},
+		})
+	}
+
+	return MessageCard{
+		Context:     context,
+		MessageType: messageType,
+		Summary:     summary,
+		ThemeColor:  getColourForState(events.GetSubjectState()),
+		Title:       title,
+		Sections: []Section{
+			{
+				ActivityTitle: activityTitleText,
+				ActivityText:  triggerDescription,
+				Facts:         facts,
+			},
+		},
+		PotentialAction: actions,
+	}
+}
+
+func (sender *Sender) buildRequest(events moira.NotificationEvents, contact moira.ContactData, trigger moira.TriggerData, plot []byte, throttled bool) (*http.Request, error) {
+
+	messageCard := sender.buildMessage(events, trigger, throttled)
+	requestURL := contact.Value
+	requestBody, err := json.Marshal(messageCard)
+	if err != nil {
+		return nil, err
+	}
+
+	sender.logger.Debugf("%s\n", requestBody)
```

Comment:
Should this debug messages appear here? If they should then we need to make them more verbose and add at least a bit of context instead of just plain JSON.

Reply:
I think we can forego this debug message in favour of the one at L161

`sender.logger.Debugf("%s %s '%s'", request.Method, request.URL.String(), string(requestBody))`

Reply:
and maybe augment it a bit to indicate that it's an MS Teams message

Reply:
Yeah, I think it is good point